multicluster/service-mirror/cluster_watcher.go

```diff
@@ -815,3 +906,331 @@ func (rcsw *RemoteClusterServiceWatcher) createOrUpdateEndpoints(ctx context.Con
+	if err != nil {
+		return nil, err
+	}
+
+	hostAnnotations := map[string]string{
+		// needed to detect real changes
+		consts.RemoteResourceVersionAnnotation: targetRef.ResourceVersion,
+		consts.RemoteServiceFqName:             fmt.Sprintf("%s.%s.%s.svc.%s", targetRef.Name, remoteService.Name, remoteService.Namespace, rcsw.link.TargetClusterDomain),
+	}
+
+	hostLabels := rcsw.getMirroredServiceLabels()
+	localServiceName := rcsw.mirroredResourceName(remoteService.Name)
+	hostLabels[consts.MirroredRootHeadlessLabel] = localServiceName
+
+	// Create service spec, clusterIP
+	hostSvc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        refLocalName,
+			Namespace:   remoteService.Namespace,
+			Annotations: hostAnnotations,
+			Labels:      hostLabels,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: remapRemoteServicePorts(remoteService.Spec.Ports),
+		},
+	}
+	hostEndpoints := &corev1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      hostSvc.Name,
+			Namespace: hostSvc.Namespace,
+			Labels: map[string]string{
+				consts.MirroredResourceLabel:  "true",
+				consts.RemoteClusterNameLabel: rcsw.link.TargetClusterName,
+			},
+			Annotations: map[string]string{
+				consts.RemoteServiceFqName: hostSvc.Annotations[consts.RemoteServiceFqName],
+			},
+		},
+		Subsets: []corev1.EndpointSubset{
+			{
+				Addresses: gatewayAddresses,
+				Ports:     rcsw.getEndpointsPorts(remoteService),
+			},
+		},
+	}
+
+	if rcsw.link.GatewayIdentity != "" {
+		hostEndpoints.Annotations[consts.RemoteGatewayIdentity] = rcsw.link.GatewayIdentity
+	}
+
+	serviceInfo := fmt.Sprintf("%s/%s", remoteService.Namespace, remoteService.Name)
+	podInfo := fmt.Sprintf("%s/%s", remoteService.Namespace, refLocalName)
+	rcsw.log.Infof("Creating new service %s for headless mirror %s", podInfo, serviceInfo)
+	createdService, err := rcsw.localAPIClient.Client.CoreV1().Services(remoteService.Namespace).Create(ctx, hostSvc, metav1.CreateOptions{})
+	if err != nil {
+		if !kerrors.IsAlreadyExists(err) {
+			// we might have created it during earlier attempt, if that is not the case, we retry
+			return createdService, RetryableError{[]error{err}}
+		}
+	}
+
+	rcsw.log.Infof("Creating a new Endpoints for %s", podInfo)
+	if _, err := rcsw.localAPIClient.Client.CoreV1().Endpoints(remoteService.Namespace).Create(ctx, hostEndpoints, metav1.CreateOptions{}); err != nil {
+		// we clean up after ourselves
+		rcsw.localAPIClient.Client.CoreV1().Services(remoteService.Namespace).Delete(ctx, refLocalName, metav1.DeleteOptions{})
+		// and retry
+		return createdService, RetryableError{[]error{err}}
+	}
+
+	return createdService, nil
+}
+
+// checkEndpointsValid looks at an Endpoints object and determines whether it is
+// valid for processing. To be valid for processing, an Endpoints object must
+// have at least one subset and at least one ready address. Without a valid
+// subset or a ready addresses we risk nil pointer dereferences (e.g looping
```

Comment:
Can you point to where we hit a nil pointer?  I would expect trying to loop through an empty or nil slice shouldn't cause any problems.  e.g. https://play.golang.org/p/mKoXlpgp2Ij